src/main/java/net/neoforged/neoforge/transfer/handlers/wrappers/items/ItemContainerContentsResourceHandler.java

```diff
@@ -0,0 +1,169 @@
+    }
+
+    @Override
+    public int size() {
+        return size;
+    }
+
+    @Override
+    public ItemResource getResource(int index) {
+        // You may need to still add what you had before. Depends on if this really resulted in much better
+        //return getContents().getImmutableStackInSlot(index).resource();
+        return ItemResource.of(getContents().getStackInSlot(index));
+    }
+
+    @Override
+    public int getAmount(int index) {
+        return getContents().getStackInSlot(index).getCount();
+    }
+
+    @Override
+    public int getCapacity(int index, ItemResource resource) {
+        Objects.checkIndex(index, size());
+        if (resource.isEmpty()) return Item.ABSOLUTE_MAX_STACK_SIZE;
+        return Math.min(resource.getMaxStackSize(), Item.ABSOLUTE_MAX_STACK_SIZE);
+    }
+
+    @Override
+    public boolean isValid(int index, ItemResource resource) {
+        ItemResource current = getResource(index);
+        return current.isEmpty() || current.equals(resource);
+    }
+
+    @Override
+    public boolean supportsInsertion(int index) {
+        return true;
+    }
+
+    @Override
+    public boolean supportsExtraction(int index) {
+        return true;
+    }
+
+    //TODO the resource should be in theory ignored here, but I want to be sure.
+    public void set(int index, ItemResource resource, int amount) {
+        ItemContainerContents contents = getContents();
+        contents.getStackInSlot(index).setCount(amount);
+        try (Transaction tx = TransactionManager.open(null)) {
+            setAndValidate(contents, amount, tx);
+            tx.commit();
+        }
+    }
+
+    @Override
+    public int insert(int index, ItemResource resource, int amount, TransactionContext context) {
+        if (ResourceHandlerUtil.isEmpty(resource, amount) || !isValid(index, resource)) return 0;
+        ItemContainerContents contents = getContents();
+        ItemStack stack = contents.getStackInSlot(index);
+        if (stack.isEmpty()) {
+            amount = Math.min(amount, resource.getMaxStackSize());
+
+            return setAndValidate(contents.with(index, resource, amount), amount, context);
+        } else if (resource.is(stack) && stack.getCount() < resource.getMaxStackSize()) {
+            int newAmount = Math.min(stack.getCount() + amount, resource.getMaxStackSize());
+            amount = newAmount - stack.getCount();
+//            if (action.isExecuting())
+            stack.grow(amount);
+            return setAndValidate(contents, amount, context);
+        }
+        return 0;
+    }
+
+    @Override
+    public int insert(ItemResource resource, int amount, TransactionContext context) {
+        if (ResourceHandlerUtil.isEmpty(resource, amount)) return 0;
+        ItemContainerContents contents = getContents();
+        int remaining = amount;
+        for (int i = 0; i < size; i++) {
```

Comment:
Please add comments about what this is doing. (My guess before reading the impl is this for loop inserts into matching stacks, skipping empty slots, and then the next for loop inserts into empty slots.)